specs/phase1/fraud-proofs.md

```diff
@@ -29,42 +32,264 @@ This document describes the shard transition function and fraud proofs as part o
+                                 slot: Slot,
+                                 shard: Shard,
+                                 shard_blocks: Sequence[SignedShardBlock],
+                                 validate_result: bool=True) -> Sequence[SignedShardBlock]:
+    choices = []
+    beacon_parent_root = get_block_root_at_slot(beacon_state, get_previous_slot(beacon_state.slot))
+    proposer_index = get_shard_proposer_index(beacon_state, slot, shard)
+    shard_blocks_at_slot = [block for block in shard_blocks if block.message.slot == slot]
+    for block in shard_blocks_at_slot:
+        temp_shard_state = shard_state.copy()  # Not doing the actual state updates here.
+        # Try to apply state transition to temp_shard_state.
+        try:
+            # Verify the proposer_index and signature
+            assert block.message.proposer_index == proposer_index
+            if validate_result:
+                assert verify_shard_block_signature(beacon_state, block)
+
+            shard_state_transition(
+                beacon_state=beacon_state,
+                shard=shard,
+                slot=slot,
+                shard_state=temp_shard_state,
+                beacon_parent_root=beacon_parent_root,
+                signed_block=block,
+            )
+        except Exception:
+            pass  # TODO: throw error in the test helper
+        else:
+            choices.append(block)
+    return choices
+```
+
+```python
+def get_proposal_at_slot(beacon_state: BeaconState,
+                         shard_state: ShardState,
+                         slot: Shard,
+                         shard: Shard,
+                         shard_parent_root: Root,
+                         shard_blocks: Sequence[SignedShardBlock],
+                         validate_result: bool=True) -> Tuple[SignedShardBlock, ShardState, Root]:
+    beacon_parent_root = get_block_root_at_slot(beacon_state, get_previous_slot(beacon_state.slot))
+    proposer_index = get_shard_proposer_index(beacon_state, slot, shard)
+    shard_state = shard_state.copy()  # Don't update the given shard_state
+    choices = get_proposal_choices_at_slot(
+        beacon_state=beacon_state,
+        shard_state=shard_state,
+        slot=slot,
+        shard=shard,
+        shard_blocks=shard_blocks,
+        validate_result=validate_result,
+    )
+    if len(choices) == 0:
+        block_header = get_empty_body_block(
+            shard_parent_root=shard_parent_root,
+            beacon_parent_root=beacon_parent_root,
+            slot=slot,
+            proposer_index=proposer_index,
+        )
+        proposal = SignedShardBlock(message=block_header)
+    elif len(choices) == 1:
+        proposal = choices[0]
+    else:
+        proposal = get_winning_proposal(beacon_state, choices)
+
+    shard_parent_root = hash_tree_root(proposal.message)
+
+    if not is_empty_body(proposal.message):
+        # Apply state transition to shard_state.
+        shard_state_transition(
```

Comment:
We might need to execute this transition because an empty block should have an affect on the gas limit (i believe).